### PR TITLE
[SPARK-8974]  Catch exceptions in allocation schedule task.

### PR DESCRIPTION
**Pull request by KaiXinXiaoLei**
I meet a problem. When I submit some tasks, the thread spark-dynamic-executor-allocation should seed the message about "requestTotalExecutors", and the new executor should start. But I meet a problem about this thread, like:

2015-07-14 19:02:17,461 | WARN  | [spark-dynamic-executor-allocation] | Error sending message [message = RequestExecutors(1)] in 1 attempts 
java.util.concurrent.TimeoutException: Futures timed out after [120 seconds]
        at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:219)
        at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)
        at scala.concurrent.Await$$anonfun$result$1.apply(package.scala:107)
        at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
        at scala.concurrent.Await$.result(package.scala:107)
        at org.apache.spark.rpc.RpcEndpointRef.askWithRetry(RpcEndpointRef.scala:102)
        at org.apache.spark.rpc.RpcEndpointRef.askWithRetry(RpcEndpointRef.scala:78)
        at org.apache.spark.scheduler.cluster.YarnSchedulerBackend.doRequestTotalExecutors(YarnSchedulerBackend.scala:57)
        at org.apache.spark.scheduler.cluster.CoarseGrainedSchedulerBackend.requestTotalExecutors(CoarseGrainedSchedulerBackend.scala:351)
        at org.apache.spark.SparkContext.requestTotalExecutors(SparkContext.scala:1382)
        at org.apache.spark.ExecutorAllocationManager.addExecutors(ExecutorAllocationManager.scala:343)
        at org.apache.spark.ExecutorAllocationManager.updateAndSyncNumExecutorsTarget(ExecutorAllocationManager.scala:295)
        at org.apache.spark.ExecutorAllocationManager.org$apache$spark$ExecutorAllocationManager$$schedule(ExecutorAllocationManager.scala:248)

when after some minutes, I find a new ApplicationMaster start,  and tasks submitted start to run. The tasks Completed. And after long time (eg, ten minutes), the number of executor  does not reduce to zero.  I use the default value of "spark.dynamicAllocation.minExecutors".

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 1]
```diff
@@ -211,7 +212,16 @@ private[spark] class ExecutorAllocationManager(
     listenerBus.addListener(listener)
 
     val scheduleTask = new Runnable() {
-      override def run(): Unit = Utils.logUncaughtExceptions(schedule())
+      override def run(): Unit = {
```
Comment (SaintBacchus) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: It's all the same code.

**Comment by KaiXinXiaoLei** [comment]
There is not same code. If there is exception in the thread of spark-dynamic-executor-allocation, the error log will be printed. but this thread does not throw exception, so this thread still support executor allocation.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 1 above)
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: The thread stops either way. How does this help?

Comment (KaiXinXiaoLei) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: If the thread not throws exception, it does not stop.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Right, but, according to the JIRA you are dealing with an exception case. How does this stop the thread from stopping?

**Comment by srowen** [comment]
This does not change the behavior though. Before, the thread would not-stop if there was no exception. This doesn't seem to be what your JIRA is about though; you are talking about the exception case. But here, all exception paths lead to the `Runnable` completing. Can you make this clearer or close this PR / JIRA?

**Comment by KaiXinXiaoLei** [comment]
I mean: If the thread throws exception, it will stop. And it will not recovery. So when the new ApplicationMaster starts and not tasks are running, the number of executor is not zero.

For example:
  1.The old ApplicationMaster is dead or disconnected
  2. Submit tasks before the new ApplicationMaster starts, 
  3. The spark-dynamic-executor-allocation thread will ask to add executors. This thread throws exception and retry three times. Then this thread stops.
  4. When the new ApplicationMaster starts, and submit tasks.
  5. After the tasks is over. the number of executors is not zero.

**Comment by srowen** [comment]
@KaiXinXiaoLei again, read your code and trace what happens when an exception is thrown from `schedule()`. The thread `run()` method either throws an exception, or completes. It stops in either case, like before. So I don't see why you believe this changes the behavior.

**Comment by srowen** [comment]
It is not true that 0 tasks means 0 executors, even with dynamic allocation. Consider a minimum > 0, or ramp-down time, or keeping executors because of cached data.

**Comment by KaiXinXiaoLei** [comment]
I set the value of "spark.dynamicAllocation.minExecutors" is 0.

**Comment by srowen** [comment]
I still don't see why this relates to the problem you describe; see my question above. If we can't make progress discussing the problem, I'd ask that you close the PR.

**Comment by vanzin** [comment]
I think the bug / PR title are just confusingly worded, but the fix is correct. It's because of this note in the `ScheduledExecutorService::scheduleAtFixedRate` documentation:

    If any execution of the task encounters an exception, subsequent executions are suppressed.

So exceptions need to be ignored while the allocation manager is running. `logUncaughtExceptions` just logs them, but lets the exceptions bubble up (and thus stop the task).

Only thing is that I never know which exception to catch in Scala (is it `ControlThrowable` or `NonFatal(e)`?).

I'd use a clearer title for the PR, something like: "Catch exceptions in allocation schedule task." You could explain why that needs to be done in the description, instead of trying to use the title for that.

**Comment by srowen** [comment]
Oh I get it, I'm wrong then. This is not about `Runnable.run()` stopping but the (Java) `Executor` stopping the scheduling of the task. I didn't realize an exception in a task cancelled the future schedule, so, now it makes sense that it does matter how `run()` finishes, not just that it finishes. I'm sorry about that, I see it now.

Can I make a minor suggestion then: change `logError` to `logWarning` since this is a recoverable error, apparently.

**Comment by KaiXinXiaoLei** [comment]
@vanzin I am sorry my description  is not clear. Thanks.
@srowen I have changed logError to logWarning. Thanks.

**Comment by srowen** [comment]
Merged into master and 1.4

**Pull request closed without merging**